SQLAnywhere corrections for DSN generation and persistent connections

andipohl opened the pull request:
I would use persistent connections. But the check with get_resource_type failed because of different strings I get for persistent and non-persistent connections.  
- persistent=false: get_resource_type() == 'SQLAnywhere connection'  
- persistent=true: get_resource_type() == 'SQLAnywhere persistent connection '  

buildDSN could not be used for mirrored database installations. Here the host-parameter has to be a list of ip-adresses, comma-separated, e.g. host=ip1:port,ip2:port 
other possibility is to leave host blank and use  

```
$driverOptions['links'] = 'tcpip(BROADCAST=YES)'
```

Please merge this changes.

Commit:
if $persistent=true, resource_type == 'SQLAnywhere persistent connnection ',
if $persistent=false, resource_type == 'SQLAnywhere connection'

Commit:
build a correct connection string for mirrored databases, too. In this case "host=ip1:port,ip2:port"
servername and dbname are optional but can also be used without host parameter

deeky666 commented:
@andipohl the issue is valid but you are mixing in `charset` option support in here which should be a separate PR. Please remove that as well as commented lines. Thanks.

deeky666 commented:
Also missing a functional test.

deeky666 commented on lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereConnection.php:
```diff
@@ -50,7 +50,7 @@ public function __construct($dsn, $persistent = false)
         $this->connection = $persistent ? @sasql_pconnect($dsn) : @sasql_connect($dsn);
 
-        if ( ! is_resource($this->connection) || get_resource_type($this->connection) !== 'SQLAnywhere connection') {
+        if ( ! is_resource($this->connection) || !preg_match('/^SQLAnywhere.*connection.*/',get_resource_type($this->connection)) ) {
```
You should change this as follows:

```php
$allowedResourceTypes = array(
    'SQLAnywhere connection',
    'SQLAnywhere persistent connection',
);
$resourceType = get_resource_type($this->connection);

if ( ! is_resource($this->connection) || in_array($resourceType, $allowedResourceTypes, true)) {
    // ...
}
```

deeky666 commented on lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php:
```diff
@@ -74,7 +74,7 @@ class SQLAnywhereStatement implements IteratorAggregate, Statement
     public function __construct($conn, $sql)
     {
-        if ( ! is_resource($conn) || get_resource_type($conn) !== 'SQLAnywhere connection') {
+        if ( ! is_resource($conn) || !preg_match('/^SQLAnywhere.*connection.*/',get_resource_type($conn))) {
```
See above.

Commit:
Revert "build a correct connection string for mirrored databases, too. In this case "host=ip1:port,ip2:port" servername and dbname are optional but can also be used without host parameter"

This reverts commit 0698cba6461ca5a641918c40efa33b8a40fb9fbc.

Commit:
Corrected PR for persistent SQLAny-Connections

* removed "charset" patch
* implemeted resource_type compare with in_array instead of regex
* implemented functional tests for connection and statement.

deeky666 commented on lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereConnection.php:
```diff
@@ -48,9 +48,16 @@ class SQLAnywhereConnection implements Connection, ServerInfoAwareConnection
-        if ( ! is_resource($this->connection) || get_resource_type($this->connection) !== 'SQLAnywhere connection') {
+        $resourceType = get_resource_type($this->connection);
+
+        if ( ! is_resource($this->connection) || !in_array($resourceType, $allowedResourceTypes, true)) {
```
TBH I think we might safely remove the resource type check at all. We should just leave the `is_resource()` check, everything else is driver specific implementation detail.

deeky666 commented on lib/Doctrine/DBAL/Driver/SQLAnywhere/SQLAnywhereStatement.php:
```diff
@@ -74,7 +74,14 @@ class SQLAnywhereStatement implements IteratorAggregate, Statement
+
+        $resourceType = get_resource_type($conn);
+
+        if ( ! is_resource($conn) || !in_array($resourceType, $allowedResourceTypes, true)) {
```
Same here

deeky666 commented on tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/ConnectionTest.php:
```diff
@@ -0,0 +1,49 @@
+        }
+    }
+
+    protected function tearDown()
```
this method can be removed

deeky666 commented on tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    protected function tearDown()
```
this method can be removed

deeky666 commented on tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/StatementTest.php:
```diff
@@ -0,0 +1,56 @@
+<?php
+namespace Doctrine\Tests\DBAL\Functional\Driver\SQLAnywhere;
```
Please add a new line before

deeky666 commented on tests/Doctrine/Tests/DBAL/Functional/Driver/SQLAnywhere/ConnectionTest.php:
```diff
@@ -0,0 +1,49 @@
+<?php
+namespace Doctrine\Tests\DBAL\Functional\Driver\SQLAnywhere;
```
Please add a new line before

Commit:
corrected functional tests

removed unnecessary  setup/teardown-methods

Commit:
remove resource type checks

I am also of the opinion that the resource type checks are unhappy. We depend on driver internals.

Ocramius commented:
I'm going to keep far away from persistent connections myself, therefore not merging into 2.5.5 myself. @deeky666 feel free to re-assign a milestone when you checked this in more depth.

andipohl commented:
What is the reason for not taking the patch in last/next release?

kimhemsoe commented:
@deeky666 is our sqlanywhere guy and dbal maintainer but he is a bit busy these days.

On that node persistent connections sounds very scary in my ears. Do there exist something like mysqlproxy and pgbouncer for sqlanywhere? If yes i would recommend to take a look at those instead.

deeky666 commented:
@andipohl sorry for the delay but thank you!

deeky666 commented:
Backported to `2.5` via https://github.com/doctrine/dbal/commit/a2d2162514a32d9473ee447f5f293eb2435f2e2b, https://github.com/doctrine/dbal/commit/eff9d44aaa3abb7ce9e7efa6598468f7209d7e54, https://github.com/doctrine/dbal/commit/da64aff62d3bf8820384553a805035eb2fbebe57, https://github.com/doctrine/dbal/commit/05c4d3fb4c724c0953262418fb2188004f450f7c, https://github.com/doctrine/dbal/commit/205617173ffab8b6a9dffc31745b61edf2c3c5cb, https://github.com/doctrine/dbal/commit/afcdacf3e382ec2c154408c015fcdb1848385f47

morozov commented:
Seems like this one introduced [buld failure on SQL Anywhere](https://continuousphp.com/git-hub/doctrine/dbal/build/34a0c230-142e-4e25-b1b7-86300c900152/output).

deeky666 commented:
Dammit, we need to check for the driver in the setup. I will send a PR. Thanks.
